closed/make/Docs.gmk

```diff
@@ -20,4 +20,115 @@
+# Text for openj9 modules
+
+OPENJ9_JAVADOC_BASE_URL := https://www.eclipse.org/openj9/
+OPENJ9_BUG_SUBMIT_URL := https://github.com/eclipse/openj9/issues
+OPENJ9_COPYRIGHT_URL :=
+OPENJ9_LICENSE_URL :=
+OPENJ9_REDISTRIBUTION_URL :=
+OPENJ9_FULL_COMPANY_NAME := IBM Corp
+OPENJ9_COMPANY_ADDRESS :=
+OPENJ9_JCL_TOPDIR=$(SUPPORT_OUTPUTDIR)/j9jcl_sources
+OPENJ9_JAVADOC_BOTTOM := \
+    <a href="$(OPENJ9_JAVADOC_BASE_URL)">Link</a> to Eclipse OpenJ9 documentation.<br> \
+    To raise a bug report or suggest an improvement create an <a href="$(OPENJ9_BUG_SUBMIT_URL)">Eclipse Openj9 issue.</a><br> \
+    Copyright &copy; 1998, $(COPYRIGHT_YEAR), $(OPENJ9_FULL_COMPANY_NAME). All rights reserved. \
+    $(DRAFT_MARKER_STR) <!-- Version $(VERSION_STRING) -->
+OPENJ9_SUBST_COMMAND := sed -i 's|JAVADOC_BOTTOM|$(OPENJ9_JAVADOC_BOTTOM)|g'
+
+# Escape any characters which have special meaning to sed
+OPENJ9_SUBST_COMMAND := $(subst &,\&,$(OPENJ9_SUBST_COMMAND))
+OPENJDK_SUBST_COMMAND := $(subst &,\&,$(OPENJDK_SUBST_COMMAND))
+
+################################################################################
+# Define a javadoc Group for the OpenJ9 specific modules
+# Module names are added to DOCS_MODULES in closed/make/common/Modules.gmk
+OpenJ9_GROUP_NAME := OpenJ9
+OpenJ9_GROUP_MODULES := openj9.*
+OpenJ9_GROUP_DESCRIPTION := \
+    Modules whose names start with {@code openj9}. contain APIs provided by the \
+    <a href="$(OPENJ9_JAVADOC_BASE_URL)">Eclipse OpenJ9</a> project. \
+    These APIs can only be used with a JDK which includes the OpenJ9 virtual machine. \
+    #
+JDK_GROUPS += OpenJ9
+
+################################################################################
+# In the jdk.management module:
+# 1. Some openj9 specific packages are added
+# 2. Some classes in the com.sun.management package are reimplemented in an
+#    openj9 specific package
+# 3. Some classes in the com.sun.management package are not relevant (cannot be
+#    used) with an openj9 based jdk.
+# The code below renames the files which cannot be used while javadoc is
+# executed so no javadoc is generated for those classes.
+#
+# This is not done within a target because the javadoc dependencies are generated
+# from the source tree before executing any targets, so dependencies would be generated
+# for files found before they are renamed, causing build errors.  
+
+OPENJDK_SUN_MGMT_SRCDIR = $(OPENJDK_JCL_TOPDIR)/jdk.management/share/classes/jdk.management/com/sun/management
+OPENJ9_SUN_MGMT_SRCDIR = $(OPENJ9_JCL_TOPDIR)/jdk.management/share/classes/com/sun/management
+
+JDK_API_CUSTOM_TARGETS += openj9-docs-set-javadoc-bottom openj9-docs-restore-files
+
+# Code should only be executed when generating the docs-jdk-api-javadoc target.
+# If executed for any other targets the renames would be done but the target
+# to restore the files would not be executed.
+
+ifneq (,$(findstring docs-jdk-api-javadoc,$(MAKECMDGOALS)))
```

Comment:
What about other targets of `Docs.gmk`, like these:
```
docs-javase-api-javadoc
docs-javase-api-modulegraph
docs-jdk-api-modulegraph
docs-jdk-index
docs-jdk-specs
docs-reference-api-javadoc
docs-reference-api-modulegraph
docs-zip
```

Reply:
Will do more testing

Reply:
All the targets run OK for me (make xxxx) except for docs-reference-api-javadoc which puts out some warnings like
`
Generating REFERENCE_API javadoc for 21 modules
javadoc: warning - bad source file: /root/openj9-openjdk-jdk11/build/linux-x86_64-normal-server-release/support/j9jcl_sources/java.base/share/classes/java/lang/LiveStackFrame.java
  file does not contain class java.lang.LiveStackFrame
  Please remove or make sure it appears in the correct subdirectory of the sourcepath.
javadoc: warning - bad source file: /root/openj9-openjdk-jdk11/build/linux-x86_64-normal-server-release/support/j9jcl_sources/java.base/share/classes/java/lang/StackFrameInfo.java
  file does not contain class java.lang.StackFrameInfo
  Please remove or make sure it appears in the correct subdirectory of the sourcepath.
`
for each of the files overridden in java.base, and then it hangs.  I don't think that is new with the changes I've made here but will investigate further.  I'm building using a docker image from the instructions here: https://www.eclipse.org/openj9/oj9_build.html, though my openj9 code needs bringing up to date.

Reply:
I've pushed the changes I've made in response to the comments so far - please could you try the build again?
`make all`, `make docs` and `make docs-jdk-api-javadoc` should all generate the javadoc (the target for the generation in the api/index.html file, so if it already exists it will not regenerate).  Remove images/docs to force javadoc to run.
Note that at the moment modules `jdk.jfr, jdk.jstatd, jdk.management.jfr` are excluded from the build (and therefore also from the javadoc generation) - also openj9.dtfjview is still missing.

Reply:
I've got another build underway for `make all`. If that works it's a sign of some missing dependencies in the makefiles - it shouldn't matter what I build before trying to build the docs.

Reply:
This looks much like what happened in my first attempt yesterday:
```
find: ‘/home/keithc/space/jdk11/build/linux-x86_64-normal-server-release/support/src/src/jdk.management/share/classes/jdk.management/com/sun/management’: No such file or directory
find: ‘/home/keithc/space/jdk11/build/linux-x86_64-normal-server-release/support/src/src/jdk.management/share/classes/jdk.management/com/sun/management’: No such file or directory
/bin/bash: line 0: cd: /home/keithc/space/jdk11/build/linux-x86_64-normal-server-release/images/docs/api: No such file or directory
/home/keithc/space/jdk11/closed/make/Docs.gmk:117: recipe for target 'openj9-docs-set-javadoc-bottom' failed
make[3]: *** [openj9-docs-set-javadoc-bottom] Error 1
make/Main.gmk:364: recipe for target 'docs-jdk-api-javadoc' failed
make[2]: *** [docs-jdk-api-javadoc] Error 1
make[2]: *** Waiting for unfinished jobs....
```

Reply:
`make JOBS=1 all` succeeded: that suggest we're just missing some dependencies.

Reply:
I think I've now fixed the dependencies and done much testing with "make JOBS=16 all" locally.

Reply:
Looks like I spoke too soon - still getting some 'directory not found' errors running the build on other platforms.  Investigating......

Reply:
These changes are now pushed to this PR, and the jdk and javadoc has been built successfully locally on x86-64_linux, ppc64le_linux, ppc64_aix, s390x_linux, x86-64_mac and x86-64_windows.